kernel/src/expressions/scalars.rs

```diff
@@ -33,8 +33,8 @@ pub enum Scalar {
 }
 
 impl Scalar {
-    pub fn data_type(&self) -> DataType {
-        match self {
+    pub fn data_type(&self) -> DeltaResult<DataType> {
```

Comment:
I'm not a huge fan of having this be fallible. It doesn't really feel like asking a scalar what its type is should be able to fail.

I get that maybe you could say you're a `Decimal` that's out of range, but at the moment we only construct it in `parse_decimal`, so if we validate there instead (which already returns a `Result`, we can just assume it's always within range here (maybe just panic if it's not to prevent weird bugs)